### ISSUE TITLE
foreign_key being VARCHAR means that an INDEX isn't used on that column

### ISSUE THREAD

**Issue by nebojsac** [CakeDC/tags#96]
Our `tagged` table is growing and we've added an INDEX on the `foreign_key` column, but have noticed that it wasn't used at all.

After more investigation I realized it's because the `foreign_key` is VARCHAR (a string) and the query generated by CakePHP uses ints. Example:
```SQL
SELECT `Tag`.`id`, `Tag`.`owner`, `Tag`.`identifier`, `Tag`.`name`, `Tag`.`keyname`, `Tag`.`weight`, `Tag`.`created`, `Tag`.`modified`, `Tagged`.`id`, `Tagged`.`foreign_key`, `Tagged`.`tag_id`, `Tagged`.`model`, `Tagged`.`language`, `Tagged`.`times_tagged`, `Tagged`.`created`, `Tagged`.`modified` FROM `tags` AS `Tag` JOIN `tagged` AS `Tagged` ON (`Tagged`.`foreign_key` IN (11366363, 11366364, 11366351, 11366352) AND `Tagged`.`tag_id` = `Tag`.`id`)  WHERE `Tagged`.`model` = 'Contact';
```
When using EXPLAIN on the query I see it's not using the INDEX on `foreign_key`. However, if I wrap the values with quote marks ( '11366363',...) then it uses it correctly.

I've added the following code to the saveTags function in the Behavior but it didn't help, it seems that CakePHP detects them as INTs:

```php
if(is_array($foreignKey))
	$foreignKey = array_map('strval',$foreignKey);
```
I've noticed that the foreign_key consistently used numbers which could easily be kept in an INT(32) column, and this might be the quickest fix for us right now. Is there a reason this would be a bad idea?

Otherwise, the fix would be changing how the 'conditions' array is generated to use multiple LIKEs.

**Comment by burzum** [comment]
> Is there a reason this would be a bad idea?

It is varchar(36) because it was build with the use of UUIDs in mind. For me it works with ints and strings in a project that uses both for the attached models. Maybe it's a DB setting? Are you using MySQL or Postgres? Might help to mention your DB.

However, this should be fixed. +1

**Comment by antoniovassell** [comment]
+1 as well, @nebojsac  as @burzum stated, stating your DB source would be great, although it may be something that is generic. In terms of fixing I guess we could check exactly where these "numbers" are generating.

**Comment by nebojsac** [comment]
I'm using MySQL. For what it's worth, I altered the table column to be INT (after confirming that it's only being filled with numbers right now) and nothing broke(yet).

After thinking about your comment @burzum I realized that VARCHAR is more flexible and is needed when developers really do use UUIDs for their primary keys. Since I only use INTs as my primary ID columns, altering the column didn't affect my codebase. So maybe it doesn't have to be "fixed" and could just be added as a *note*?

After some digging around I figured out that I was looking in the wrong file - it's not in the TaggableBehaviour. It actually happens either in Containable or in somewhere in the Core when bindModel happens (which is defined in Tagged.php, _findTagged), because I get that query when doing:
```php
$this->Model->find('all',array('contain'=>array('Tag')));
```

So it's either the bindModel, or Containable that turns the strings int integers. So this bug might be at the wrong place. I still haven't been able to pinpoint where exactly it happens though.
